### PR TITLE
Some improvements

### PR DESCRIPTION
**Pull request by Psychokiller1888**
Hi there Zurich!

This adds onMessageFrom and proper onMessage while renaming onMessage to onMessageIs. This way, Construct 2 can listen for any messages income and for messages coming in from a specific device. The old onMessage is too restrictive by forcing us to specify deviceId and message

(to airconsole 1.5.0 commit was reverted btw, adding it later, with the missing functions)

### CODE REVIEW

File: plugin/files/airconsole/runtime.js [hunk 1]
```diff
@@ -77,7 +77,9 @@ cr.plugins_.AirConsole = function(runtime)
           self.runtime.trigger(cr.plugins_.AirConsole.prototype.cnds.OnMessageKey, self);
         }
         if (data.message) {
-          self.runtime.trigger(cr.plugins_.AirConsole.prototype.cnds.OnMessage, self);
+          self.runtime.trigger(cr.plugins_.AirConsole.prototype.cnds.OnMessageIs, self);
+		  self.runtime.trigger(cr.plugins_.AirConsole.prototype.cnds.OnMessageFrom, self);
+		  self.runtime.trigger(cr.plugins_.AirConsole.prototype.cnds.OnMessage, self);
```
Comment (francois-dibulo) [inline plugin/files/airconsole/runtime.js]: When a message is received I either want to know:
- From which device and which message-data it contains (prev. OnMessage)
- Any device and which message it contains (OnMessageKey)

Like this we trigger OnMessage always.

Also, wouldn't you check anyways what the message-data is if you know that a specific device sent it?
(which in the end is OnMessage). Can you give a use-case?

(Please also check the indentation)

File: plugin/files/airconsole/runtime.js [hunk 2]
```diff
@@ -86,6 +88,7 @@ cr.plugins_.AirConsole = function(runtime)
       if (!device_data) {
         self.ac_leave_id = device_id;
         self.runtime.trigger(cr.plugins_.AirConsole.prototype.cnds.OnDeviceLeft, self);
+		self.runtime.trigger(cr.plugins_.AirConsole.prototype.cnds.OnAnyDeviceLeft, self);
```
Comment (francois-dibulo) [inline plugin/files/airconsole/runtime.js]: Indentation please :)

File: plugin/files/airconsole/edittime.js [hunk 3]
```diff
@@ -37,11 +37,11 @@ AddCmpParam("Is", "Compare message data key");
 AddAnyTypeParam("Value", "Compare to this value");
 AddCondition(0, cf_trigger, "On message key", "Data", "On message key is {2}", "Triggered when a message key is received from a device.", "OnMessageKey");
 
-// Receive onMessage
+// Receive onMessageIs
 AddAnyTypeParam("Value", "Compare to this value");
 // AND DEVICE ID IS
 AddAnyTypeParam("Object DeviceID", "The object associated to this device id");
-AddCondition(1, cf_trigger, "On message", "Data", "On message is {0} and from {1}", "Triggered when a message is received from a device.", "OnMessage");
+AddCondition(1, cf_trigger, "On specific message", "Data", "On message is {0} and from {1}", "Triggered when a specific message is received from a device.", "OnMessageIs");
```
Comment (francois-dibulo) [inline plugin/files/airconsole/edittime.js]: Is this backwards compatible?

**Review by francois-dibulo — COMMENTED** [review]
Thanks for the PR! Left some comments for clarification!

File: plugin/files/airconsole/edittime.js (hunk 3 above)
Comment (Psychokiller1888) [reply plugin/files/airconsole/edittime.js]: Yeah, this is just renaming the function, should be ok. Will try in a few minutes

File: plugin/files/airconsole/runtime.js (hunk 2 above)
Comment (Psychokiller1888) [reply plugin/files/airconsole/runtime.js]: Missed that one, sry ;) I've fixed them all for a next PR, containing missing functions and upgrade to 1.5.0

http://puu.sh/rM0An.png
http://puu.sh/rM0FN.png

File: plugin/files/airconsole/runtime.js (hunk 1 above)
Comment (Psychokiller1888) [reply plugin/files/airconsole/runtime.js]: Well, I want to know if we receive a message, from whoever it is and whatever it contains

http://puu.sh/rM0gJ.png

Original OnMessage restrics  in a way that the message reception only triggers if we know both other conditions, DeviceId and Message. With these two, we can trigger  actions if a specific device sent a message, or if anybody sent anything. Applications? Well, say I build a pictionary game, first to answer wins, my controllers will send strings back to the screen and onMessage will trigger. I can't do that with the old implementation, I need to know the message and the deviceId

File: plugin/files/airconsole/edittime.js (hunk 3 above)
Comment (Psychokiller1888) [reply plugin/files/airconsole/edittime.js]: Yes it is. Reverted back to master, created anew  project, just added OnMessage, saved, closed, reinjected the updated one, reopened and it's ok

Comment (francois-dibulo) [reply plugin/files/airconsole/edittime.js]: 👍

File: plugin/files/airconsole/runtime.js (hunk 1 above)
Comment (francois-dibulo) [reply plugin/files/airconsole/runtime.js]: Ok! Can you rephrase the descriptions to:

OnMessageIs:
  Triggered when a specific message is received from a specific device.

OnMessage:
  Triggered when any message is received from any device

OnMessageFrom:
  Triggered when any message is received from a specific device

Thanks!

File: plugin/files/airconsole/runtime.js (hunk 2 above)
Comment (francois-dibulo) [reply plugin/files/airconsole/runtime.js]: Great! Can you fix those things and then push the new changes with 1.5.0?! Saw you also implemented the highscores! 👍

Comment (Psychokiller1888) [reply plugin/files/airconsole/runtime.js]: Sure, will do. Didn't want to push the highscores, I have no way to test them really. Communication works, triggers do their job, but I don't know about the data received. Need to upload a finished game on your servers, the simulator doesn't seem to accept highscores. Can push as an experimental implementation, marked as deprecated in C2

**Comment by Psychokiller1888** [comment]
Ok, this is my current state

I have one problem though and I suspect C2 to be in fault. When using 1.5.0, we get very often a js error when previewing. Instantiation fails on new AirConsole(). The error usually disappears on a simple F5... The error never happens when exporting the project to html or whatever. So delay problems maybe? Or the dynamic <script> tag generation? I simply don't get it... Works 9 timeson 10 for me, for others it  fails 5 times on 10...

http://puu.sh/rNpJr.png

Don't be afraid, there's more coming, I'm just overwhelmed with work these last days. Let me know if you're nok with my changes
